### PR TITLE
implement NaCl - phase 2 - integrate TPM keys

### PR DESCRIPTION
**Pull request by James-Pickett**
This PR add support for TPM keyer to leverage TPM hardware for crypto operations. Test in CI are done using a TPM simulator. Makes needed adjustments to allow tests to run on windows and macos.

We are moving krypto toward implementing [NaCl](https://doc.libsodium.org/) since it simplifies the boxing process and there is less room for error. However, in order to have future compatibility with MacOS secure enclave, we are pre-computing P256 elliptic curve keys and passing them to NaCl rather than using the native NaCl ed25519 elliptic curve keys.

Below is a rough outline of how I plan to move forward with this and I'll do my best to keep PRs smallish.

- [x] phase 1 - basic cross language NaCl seal / open
- [ ] phase 2 - **integrate TPM keys  < -- this PR**
- [ ] phase 3 - encode sealed NaCl boxes into pngs
- [ ] phase 4 - add sign / verify ?
- [ ] phase 5 - integrate Secure Enclave ?

### CODE REVIEW

File: cross_language_tests/nacler_cross_test.go [hunk 1]
```diff
@@ -50,6 +50,9 @@ func TestNaclerRuby(t *testing.T) {
 	bobRubyKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
 	require.NoError(t, err)
 
+	malloryKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	require.NoError(t, err)
+
 	naclerTests := []struct {
```
Comment (James-Pickett) [inline cross_language_tests/nacler_cross_test.go:56]: Wonder if we should add a TPM keyer to these tests? There is a TPM keyer test in `pkg/nacler` that test the tpm keys against the go local keys. This tests the go local keys against ruby local keys, by transitive property, tpm keys will work against ruby local keys?

File: pkg/nacler/keyers/tpmkeyer/tpmkeyer.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
+package tpmkeyer
```
Comment (James-Pickett) [inline pkg/nacler/keyers/tpmkeyer/tpmkeyer.go:1]: not convinced on package name .... it's a bit redundant `pkg/nacler/keyers/tpmkeyer` ... but `pkg/nacler/keyers/tpm` felt weird using in code, for example calling `tpm.New()`

File: pkg/nacler/keyers/tpmkeyer/tpmkeyer.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+type TpmKeyerOption func(*TpmKeyer)
+
+// WithExternalTpm lets the caller provide a tpm to use. This is useful for testing.
+// The caller is responsible for closing the external tpm.
+func WithExternalTpm(externalTpm io.ReadWriteCloser) TpmKeyerOption {
+	return func(t *TpmKeyer) {
+		t.externalTpm = externalTpm
+	}
+}
```
Comment (directionless) [inline pkg/nacler/keyers/tpmkeyer/tpmkeyer.go:25]: Based mostly off this comment, this seems a little weird. Why have a single `TpmKeyer` with internal switching between an internal and external, and not have `InternalTpmKeyer` and `ExternalTpmKeyer` (or whatever)

File: pkg/nacler/nacler.go [hunk 4]
```diff
@@ -54,3 +58,21 @@ func (n *Nacler) Open(cipherText []byte) ([]byte, error) {
 	}
 	return decrypted, nil
 }
+
+func (n *Nacler) cachedSharedKey() ([32]byte, error) {
```
Comment (directionless) [inline pkg/nacler/nacler.go:62]: Does it make sense to call this in `New` instead of inline with Seal or Open?

**Review by directionless — COMMENTED** [review]
Generally looks really straight forward. I think I need to work through the actual crypto and see if it conforms to examples, I don't know that code path.

I'm also unsure about the whole internal vs external

File: pkg/nacler/keyers/tpmkeyer/tpmkeyer.go (hunk 2 above)
Comment (directionless) [reply pkg/nacler/keyers/tpmkeyer/tpmkeyer.go:1]: I think the alternative, would be to call it `pkg/nacler/keyers/tpm`, and if places needed to import both, they'd tpmkeyer "pkg/nacler/keyers/tpm"`.

But this seemed fine to me

File: cross_language_tests/nacler_cross_test.go (hunk 1 above)
Comment (James-Pickett) [reply cross_language_tests/nacler_cross_test.go]: I added the tpm to the cross tests ... it was painful due to the ruby install action adding an incompatible version of gcc to the path, but got it figured out.

File: pkg/nacler/keyers/tpmkeyer/tpmkeyer.go (hunk 2 above)
Comment (James-Pickett) [reply pkg/nacler/keyers/tpmkeyer/tpmkeyer.go]: renamed package to just tpm

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -1,61 +1,78 @@
...
     - name: Set up Ruby
       uses: ruby/setup-ruby@v1
       with:
         ruby-version: ${{ matrix.ruby }}
         bundler-cache: true
+    - name: Install Libsodium on Windows
+      if: runner.os == 'Windows'
+      run: |
+        cd $env:temp
+        Invoke-WebRequest "https://download.libsodium.org/libsodium/releases/libsodium-1.0.18-stable-msvc.zip" -OutFile "./libsodium.zip"
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply .github/workflows/test.yml:33]: just ruby added comment to action yml

File: pkg/nacler/keyers/tpmkeyer/tpmkeyer.go (hunk 3 above)
Comment (James-Pickett) [reply pkg/nacler/keyers/tpmkeyer/tpmkeyer.go]: I updated this comment, but I'm not sure if were on the same page. The `TpmKeyer` does not change, what changes is the way we interface with the TPM chip hardware. For example, in the CI environment, we don't have access to a TPM chip, so we set an "ExternalTpm" to the tpm simulator.

File: pkg/nacler/nacler.go (hunk 4 above)
Comment (James-Pickett) [reply pkg/nacler/nacler.go]: yea, that is cleaner and then we don't need mutex

**Review by directionless — APPROVED** [review]
I think it's good! The interfaces make this look pretty straight forward.

**Pull request merged**